### ISSUE TITLE
Make libpostal_setup_* functions return an object instead of global allocation?

### ISSUE THREAD

**Issue by GuillaumeGomez** [openvenues/libpostal#475]
Currently, it's not possible to have multiple instances of libpostal in one process. Even worst, if you want to change the `datadir`, you're changing it for everything. It also brings multithreading conflicts. Wouldn't it be possible to returns objects and expect this object in all functions using this object instead?

Having an instance for each kind would prevent all the previously mentioned issues but would also make it more clear in the API which function is using which initializer.

What do you think?

**Comment by AeroXuk** [comment]
Personally I don’t think the global allocation is an issue, so long as all the other functions only use it “read only”. Otherwise you’ll be loading over 2GB of model data per instance.

**Comment by GuillaumeGomez** [comment]
It's up to the user to use it "correctly". For instance, you can "group" instances if they are created with the same parameters since they only provide read-only access. The goal is to prevent conflicts and trying to make the API a bit more clear until we have more documentation. :)

If you make it clear in the documentation that creating a new initializer is very expensive, people will be careful (hopefully). But at least they have the choice and can handle multiple instances at once if they want to.

**Comment by GuillaumeGomez** [comment]
To give an actual example of why this is an issue, I'm currently binding the library in Rust and I end up with code looking like [this](https://github.com/GuillaumeGomez/libpostal-rs/blob/master/src/language_classifier.rs#L41-L70). This is suboptimal. :-/

**Comment by GuillaumeGomez** [comment]
If this a problem of time on your side to implement it, I can try to do it? But I first need to be sure that you agree with this solution first. :)

**Comment by AeroXuk** [comment]
Whilst @albarrentine hasn’t provided any official documentation on the API, I did make a start on documenting it on the project wiki.
Everyone is free to contribute to the wiki page to help get the API documented.

https://github.com/openvenues/libpostal/wiki/API

**Comment by GuillaumeGomez** [comment]
Yes, this is what I'm currently using. This is a nice start but far from enough. ;) For example, even after looking at the source code, I have no idea what [libpostal_is_name_duplicate_fuzzy](https://github.com/openvenues/libpostal/wiki/API#libpostal_is_name_duplicate_fuzzy) `token_scores*` arguments refer to. I'd love to help on the documentation but if I don't know how to use it, I can't document it. I can help on the rest though. Will give it a try!

**Comment by GuillaumeGomez** [comment]
So anyway, back to original topic: do you think it'd be worth it to change the initializers the way I suggested?

**Comment by AeroXuk** [comment]
I’m just a fellow user/contributor so can only speak for myself.
From looking at the bug you highlighted, that is a confusing side effect if a second call to init changes the first call. So I support something to fix that.
I like I expect most people will, keep all model files in the same data dir (due to the update script working that way.

I can see a user case for being able to have multiple instances in the same process for people who might want to train custom models and benchmark them against the official model. Although I think this is also quiet an edge use case.

As long at documentation and project read me warns casual users to only initialise once it could be fine.

I would expect every function would need to change definition to take a pointer to the shared/instance data, which would probably have to be timed with a major version number update as it would be a breaking change.

I think @albarrentine would need to comment here if he is happy with this change in the next major version.

**Comment by GuillaumeGomez** [comment]
Yes, API change so it'll require a major version update. But this is a great news! Waiting for the confirmation then I'll implement it.

Thanks for answering in any case!
